### PR TITLE
Add compilation time analysis.

### PR DESCRIPTION
**Pull request by golechwierowicz**

### CODE REVIEW

File: benchmarks/aggregate.py [hunk 1]
```diff
@@ -36,6 +36,38 @@
...
+        'metrics'] else []
+    if len(total_times) <= 1:
+      return None
+    compile_and_run_time = total_times[0]
+    run_time = np.average(total_times[1:])
+    std = np.std(total_times[1:])
+    compile_time = compile_and_run_time - run_time
+    # Single sample size for compilation time, std = 0.
+    # Using stddev of the run time distribution.
+    return Datapoint(compile_time, 0)
```
Comment (golechwierowicz) [inline benchmarks/aggregate.py:52]: @ecota,

I eventually put 0 in here, but in general I'm not sure what to put here, as n = 1 for compilation estimate.

Comment (cota) [reply benchmarks/aggregate.py:52]: 0 is the only reasonable value here. stddev of the sample divides by `N-1`, so if N==1 we'd have infinite stddev. Which would be fair enough (we have no clue about the distribution with just one sample), but we're not going to plot infinite error bars :-)

File: benchmarks/aggregate.py [hunk 2]
```diff
@@ -36,6 +36,38 @@
...
+class DatapointSelector:
+
+  def compile(row):
+    total_times = row['metrics']['total_time'] if 'total_time' in row[
+        'metrics'] else []
+    if len(total_times) <= 1:
+      return None
+    compile_and_run_time = total_times[0]
+    run_time = np.average(total_times[1:])
+    std = np.std(total_times[1:])
```
Comment (cota) [inline benchmarks/aggregate.py:48]: Let's not compute std then -- AFAICT it's unused.

File: benchmarks/aggregate.py [hunk 3]
```diff
@@ -36,6 +36,38 @@
...
+    compile_time = compile_and_run_time - run_time
+    # Single sample size for compilation time, std = 0.
+    # Using stddev of the run time distribution.
```
Comment (cota) [inline benchmarks/aggregate.py:51]: This comment is also obsolete since std is set to 0.

**Comment by cota** [comment]
Other than the small nit above, this LGTM with the caveat that we want to double check that this methodology is sound.

File: benchmarks/aggregate.py [hunk 4]
```diff
@@ -90,21 +122,10 @@ def process_file(args, results_map: Dict[str, Any], filename: str):
...
         continue
-      total_times = r['metrics']['total_time'] if 'total_time' in r[
-          'metrics'] else []
-      # Skip the first three elements in total_times[]; the first one
-      # includes compilation time, the 2nd and 3rd warm up the caches.
-      skip_elems = 3
-      if len(total_times) <= skip_elems:
+
+      dp = getattr(DatapointSelector, args.metric)(r)
+      if dp is None:
```
Comment (vanbasten23) [inline benchmarks/aggregate.py:127]: is dp used if it's not None?

File: benchmarks/aggregate.py (hunk 1 above)
Comment (golechwierowicz) [reply benchmarks/aggregate.py:52]: Thanks, my reasoning was that if 100% of the data samples (n=1) have the same value then std dev has to be 0, although I felt there some holes in it.

File: benchmarks/aggregate.py (hunk 4 above)
Comment (golechwierowicz) [reply benchmarks/aggregate.py:127]: It's used on line 140.

File: benchmarks/aggregate.py (hunk 3 above)
Comment (golechwierowicz) [reply benchmarks/aggregate.py:51]: Done.

File: benchmarks/aggregate.py (hunk 2 above)
Comment (golechwierowicz) [reply benchmarks/aggregate.py:48]: Done.

**Review by cota — APPROVED** [review]

**Pull request merged**